pkg/controller/certificatesigningrequest/approving_controller.go

```diff
@@ -0,0 +1,192 @@
+	name() string
+}
+
+// CSRApprovingController is responsible for approving CertificateSigningRequests.
+type CSRApprovingController struct {
+	client            clientset.Interface
+	csrLister         csrlister.CertificateSigningRequestLister
+	csrListerSynced   cache.InformerSynced
+	autoApprovePolicy config.AutoApprovePolicy
+	queue             workqueue.RateLimitingInterface
+	approvers         []approver
+}
+
+// NewCSRApprovingController returns a new *CSRApprovingController.
+func NewCSRApprovingController(client clientset.Interface, csrInformer csrinformer.CertificateSigningRequestInformer, approvePolicy config.AutoApprovePolicy) *CSRApprovingController {
+	c := &CSRApprovingController{
+		client:            client,
+		csrLister:         csrInformer.Lister(),
+		csrListerSynced:   csrInformer.Informer().HasSynced,
+		autoApprovePolicy: approvePolicy,
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "certificateSigningRequest"),
+		approvers: []approver{
+			&ipsecCSRApprover{
+				client:        client,
+				approvePolicy: approvePolicy,
+			},
+		},
+	}
+	csrInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: c.enqueueCertificateSigningRequest,
+		},
+		resyncPeriod,
+	)
+	return c
+}
+
+// Run begins watching and syncing of the CSRApprovingController.
+func (c *CSRApprovingController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s", approverControllerName)
+	defer klog.Infof("Shutting down %s", approverControllerName)
+
+	cacheSyncs := []cache.InformerSynced{c.csrListerSynced}
+	if !cache.WaitForNamedCacheSync(approverControllerName, stopCh, cacheSyncs...) {
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *CSRApprovingController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *CSRApprovingController) enqueueCertificateSigningRequest(obj interface{}) {
+	csr, ok := obj.(*certificatesv1.CertificateSigningRequest)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Received unexpected object: %v", obj)
+			return
+		}
+		csr, ok = deletedState.Obj.(*certificatesv1.CertificateSigningRequest)
+		if !ok {
+			klog.Errorf("DeletedFinalStateUnknown contains non-CertificateSigningRequest object: %v", deletedState.Obj)
+			return
+		}
+	}
+	c.queue.Add(csr.Name)
+}
+
+func (c *CSRApprovingController) syncCSR(key string) error {
+	startTime := time.Now()
+	defer func() {
+		d := time.Since(startTime)
+		klog.V(2).Infof("Finished syncing CertificateSigningRequest %s. (%v)", key, d)
+	}()
+
+	csr, err := c.csrLister.Get(key)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	if approved, denied := getCertApprovalCondition(&csr.Status); approved || denied {
+		return nil
```

Comment:
I think it is not common to use update/patch APIs for CSRs for spec updates. It is hard to track the status of the most recent CSR spec if it can be updated. I also removed the update/patch permissions from antrea-agent for CSRs.

Reply:
Probably good to add some comments here.